Content.Server/Ninja/Systems/NinjaSuitSystem.cs

```diff
@@ -62,12 +68,39 @@ private void OnSuitInsertAttempt(EntityUid uid, NinjaSuitComponent comp, Contain
         if (!_powerCell.TryGetBatteryFromSlot(uid, out var battery))
             return;
 
+        var user = Transform(uid).ParentUid;
+
+        // as funny as the concept of inserting weapon-grade power cages into the small and nimble ninja suit is, no.
+        if (!TryComp<ItemComponent>(args.EntityUid, out var itemComp) || _item.GetSizePrototype(itemComp.Size) >= _item.GetSizePrototype(NinjaSuitInsertableCellSize)) {
+            args.Cancel();
+            Popup.PopupEntity(Loc.GetString("ninja-cell-too-large"), user, user);
+            return;
+        }
+
         // can only upgrade power cell, not swap to recharge instantly otherwise ninja could just swap batteries with flashlights in maints for easy power
-        if (!TryComp<BatteryComponent>(args.EntityUid, out var inserting) || inserting.MaxCharge <= battery.MaxCharge)
+        if (TryComp<BatteryComponent>(args.EntityUid, out var inserting)) {
+            // assigns a score to both the new and old cell equal to their maximum capacity
+            var oldscore = battery.MaxCharge;
+            var newscore = inserting.MaxCharge;
+            // if a cell is able to automatically recharge, boost the score drastically depending on the recharge rate
+            // this is to ensure a ninja can still add a micro reactor cell even if they already have a medium or high.
+            if(TryComp<BatterySelfRechargerComponent>(battery.Owner, out var oldself))
+                if(oldself.AutoRecharge)
+                    oldscore = oldscore + (oldself.AutoRechargeRate*100);
+            if(TryComp<BatterySelfRechargerComponent>(args.EntityUid, out var newself))
+                if(newself.AutoRecharge)
+                    newscore = newscore + (newself.AutoRechargeRate*100);
+            if(newscore <= oldscore) {
+                args.Cancel();
+                Popup.PopupEntity(Loc.GetString("ninja-cell-downgrade"), user, user);
+                return;
+            }
+        } else {
             args.Cancel();
+            return;
+        }
```

Comment:
// can only upgrade power cell, not swap to recharge instantly otherwise ninja could just swap batteries with flashlights in maints for easy power
if (!TryComp<BatteryComponent>(args.EntityUid, out var inserting))
{
    args.Cancel();
    return;
}

// assigns a score to both the new and old cell equal to their maximum capacity
var oldscore = battery.MaxCharge;
var newscore = inserting.MaxCharge;

// if a cell is able to automatically recharge, boost the score drastically depending on the recharge rate
// this is to ensure a ninja can still add a micro reactor cell even if they already have a medium or high.
if (TryComp<BatterySelfRechargerComponent>(uid, out var oldself) && oldself.AutoRecharge)
    oldscore += oldself.AutoRechargeRate * 100;

if (TryComp<BatterySelfRechargerComponent>(args.EntityUid, out var newself) && newself.AutoRecharge)
    newscore += newself.AutoRechargeRate * 100;

if (newscore <= oldscore)
{
    args.Cancel();
    Popup.PopupEntity(Loc.GetString("ninja-cell-downgrade"), user, user);
    return;
}

Reply:
also make the score logic a function so its not getting copy pasted

Reply:
Done